vijaynadimpalli/ps.py

```diff
@@ -0,0 +1,437 @@
+    Extended approach,Geometric approach and Extremal approach.
+    https://www.econstor.eu/bitstream/10419/147450/1/870932616.pdf
+    """
+
+    def __init__(self, df: pd.DataFrame):
+        """
+        Inputs the price series required for further calculations.
+        Also includes preprocessing steps described in the paper, before starting the Partner Selection procedures.
+        These steps include, finding the returns and ranked returns of the stocks, and calculating the top 50
+        correlated stocks for each stock in the universe.
+
+        :param df: (pd.DataFrame): Contains price series of all stocks in universe
+        """
+
+        if len(df) == 0:
+            raise Exception("Input does not contain any data")
+
+        if not isinstance(df, pd.DataFrame):
+            raise Exception("Partner Selection Class requires a pandas DataFrame as input")
+
+        self.universe = df  # Contains daily prices for all stocks in universe.
+        self.returns, self.ranked_returns = self._get_returns()  # Daily returns and corresponding ranked returns.
+
+        # Correlation matrix containing all stocks in universe
+        self.correlation_matrix = self._correlation()
+        # For each stock in universe, tickers of top 50 most correlated stocks are stored
+        self.top_50_correlations = self._top_50_tickers()
+        # Quadruple combinations for all stocks in universe
+        self.all_quadruples = self._generate_all_quadruples()
+
+    def _correlation(self) -> pd.DataFrame:
+        """
+        Calculates correlation between all stocks in universe
+
+        :return: (pd.DataFrame) : Correlation Matrix
+        """
+
+        return self.ranked_returns.corr(method='pearson')  # pearson or spearman,we get same results as input is ranked
+
+    def _get_returns(self) -> (pd.DataFrame, pd.DataFrame):
+        """
+        Calculating daily returns and ranked daily returns of the stocks
+
+        :return (tuple):
+            returns_df : (pd.DataFrame) : Dataframe consists of daily returns
+            returns_df_ranked : (pd.DataFrame) : Dataframe consists of ranked daily returns between [0,1]
+        """
+
+        returns_df = (self.universe - self.universe.shift(1))
+        returns_df = returns_df / self.universe.shift(1)
+        returns_df = returns_df.replace([np.inf, -np.inf], np.nan).ffill().dropna()
+
+        # Calculating rank of daily returns for each stock. 'first' method is used to assign ranks in order they appear
+        returns_df_ranked = returns_df.rank(axis=0, method='first', pct=True)
+        return returns_df, returns_df_ranked
+
+    def _top_50_tickers(self) -> pd.DataFrame:
+        """
+        Calculates the top 50 correlated stocks for each target stock
+
+        :return: (pd.DataFrame) : Dataframe consisting of 50 columns for each stock in the universe
+        """
+
+        def tickers_list(col):
```

Comment:
I'm personally not a fan of nested methods but if it's only used here, good design choice.

Could add the @staticmethod decorator.